File: zig/src/data_structures/hash_table.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Hash table using std.HashMap.
/// Time: O(1) average for all operations
/// Space: O(n)
pub fn HashTable(comptime K: type, comptime V: type) type {
    return struct {
        const Self = @This();
        const Map = std.AutoHashMap(K, V);

        map: Map,

        pub fn init(allocator: Allocator) Self {
            return .{
                .map = Map.init(allocator),
            };
        }

        pub fn deinit(self: *Self) void {
            self.map.deinit();
        }

        pub fn put(self: *Self, key: K, value: V) !void {
            try self.map.put(key, value);
        }

        pub fn get(self: *const Self, key: K) ?V {
            return self.map.get(key);
        }

        pub fn remove(self: *Self, key: K) ?V {
            const kv = self.map.fetchRemove(key);
            return if (kv) |entry| entry.value else null;
        }

        pub fn contains(self: *const Self, key: K) bool {
            return self.map.contains(key);
        }

        pub fn isEmpty(self: *const Self) bool {
            return self.map.count() == 0;
        }

        pub fn size(self: *const Self) usize {
            return self.map.count();
        }

        pub fn clear(self: *Self) void {
            self.map.clearRetainingCapacity();
        }

        pub fn iterator(self: *const Self) Map.Iterator {
            return self.map.iterator();
        }
    };
}

test "HashTable operations" {
    const allocator = std.testing.allocator;
    var table = HashTable(i32, i32).init(allocator);
    defer table.deinit();

    try table.put(1, 10);
    try table.put(2, 20);
    try table.put(3, 30);

    try std.testing.expectEqual(@as(?i32, 10), table.get(1));
    try std.testing.expectEqual(@as(?i32, 20), table.get(2));
    try std.testing.expectEqual(@as(?i32, null), table.get(4));
}

File: elixir/lib/data_structures/hash_table.ex
defmodule DataStructures.HashTable do
  @moduledoc """
  Hash table implementation using Elixir's Map.
  Time: O(1) average for all operations
  Space: O(n)
  """

  defstruct map: %{}

  @type t :: %__MODULE__{map: map()}

  @doc "Create an empty hash table"
  @spec new() :: t()
  def new, do: %__MODULE__{}

  @doc "Create from keyword list or map"
  @spec from_list([{any(), any()}]) :: t()
  def from_list(list) when is_list(list) do
    %__MODULE__{map: Map.new(list)}
  end

  @doc "Insert or update a key-value pair"
  @spec put(t(), any(), any()) :: t()
  def put(%__MODULE__{map: map}, key, value) do
    %__MODULE__{map: Map.put(map, key, value)}
  end

  @doc "Get value by key"
  @spec get(t(), any()) :: {:ok, any()} | {:error, :not_found}
  def get(%__MODULE__{map: map}, key) do
    case Map.fetch(map, key) do
      {:ok, value} -> {:ok, value}
      :error -> {:error, :not_found}
    end
  end

  @doc "Get value by key with default"
  @spec get(t(), any(), any()) :: any()
  def get(%__MODULE__{map: map}, key, default) do
    Map.get(map, key, default)
  end

  @doc "Remove a key-value pair"
  @spec remove(t(), any()) :: {any() | nil, t()}
  def remove(%__MODULE__{map: map}, key) do
    {Map.get(map, key), %__MODULE__{map: Map.delete(map, key)}}
  end

  @doc "Check if key exists"
  @spec contains?(t(), any()) :: boolean()
  def contains?(%__MODULE__{map: map}, key) do
    Map.has_key?(map, key)
  end

  @doc "Get all keys"
  @spec keys(t()) :: [any()]
  def keys(%__MODULE__{map: map}), do: Map.keys(map)

  @doc "Get all values"
  @spec values(t()) :: [any()]
  def values(%__MODULE__{map: map}), do: Map.values(map)

  @doc "Check if hash table is empty"
  @spec empty?(t()) :: boolean()
  def empty?(%__MODULE__{map: map}), do: map_size(map) == 0

  @doc "Get number of key-value pairs"
  @spec size(t()) :: non_neg_integer()
  def size(%__MODULE__{map: map}), do: map_size(map)

  @doc "Convert to list of key-value tuples"
  @spec to_list(t()) :: [{any(), any()}]
  def to_list(%__MODULE__{map: map}), do: Map.to_list(map)

  @doc "Clear all entries"
  @spec clear(t()) :: t()
  def clear(%__MODULE__{}), do: %__MODULE__{}
end
